### PR TITLE
Issue #5511: Allow blank lines in Header

### PR DESCRIPTION
**Pull request by kemitix**
Issue #5511

Moves the check to replace a blank line with an regex pattern from `AbstractHeaderCheck`
to `RegexpHeaderCheck`. `RegexpHeaderCheck` now also adds the blank regexp pattern into
log messages in place of a blank line to be consistent with existing behaviour (i.e. so
that existing tests in `RegexpHeaderCheckTest` continue to pass unchanged)

Includes a test to verify that a header with blank lines does not cause a violation.

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/header/RegexpHeaderCheck.java [hunk 1]
```diff
@@ -115,6 +117,20 @@ protected void processFiltered(File file, FileText fileText) {
         }
     }
 
+    /**
+     * Returns the line from the header. Where the line is blank return the regexp pattern
+     * for a blank line.
+     * @param headerLineNo heaeder line number to return
+     * @return the line from the header
+     */
+    private String header(int headerLineNo) {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/header/RegexpHeaderCheck.java]: Method name should start with a verb.
`header` => `getHeaderLine`

Comment (kemitix) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/header/RegexpHeaderCheck.java]: @rnveach Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/header/RegexpHeaderCheck.java [hunk 2]
```diff
@@ -115,6 +117,20 @@ protected void processFiltered(File file, FileText fileText) {
...
     }
 
+    /**
+     * Returns the line from the header. Where the line is blank return the regexp pattern
+     * for a blank line.
+     * @param headerLineNo heaeder line number to return
+     * @return the line from the header
+     */
+    private String getHeaderLine(int headerLineNo) {
+        String header = getHeaderLines().get(headerLineNo);
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/header/RegexpHeaderCheck.java]: please name it as `line` ,as  term header we use for block of lines.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/header/RegexpHeaderCheck.java [hunk 3]
```diff
@@ -151,9 +167,15 @@ private boolean isMultiLine(int lineNo) {
     @Override
     protected void postProcessHeaderLines() {
         final List<String> headerLines = getHeaderLines();
-        for (String line : headerLines) {
+        final Pattern blankLine = Pattern.compile(EMPTY_LINE_PATTERN);
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/header/RegexpHeaderCheck.java]: please move compilation of pattern to class field, to be done ones, for whole execution of checkstyle.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/header/RegexpHeaderCheck.java [hunk 4]
```diff
@@ -151,9 +167,15 @@ private boolean isMultiLine(int lineNo) {
...
+        for (final String line : headerLines) {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/header/RegexpHeaderCheck.java]: please remove this `final` it pollute code and we have special Check for this that should make change of value impossible.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java [hunk 5]
```diff
@@ -172,13 +172,10 @@ private void loadHeader(final Reader headerReader) throws IOException {
         final LineNumberReader lnr = new LineNumberReader(headerReader);
         try {
             while (true) {
-                String line = lnr.readLine();
+                final String line = lnr.readLine();
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java]: please refactor code a bit more like:
```
do{
 String line = lnr.readLine();
 if (line != null) {
   readerLines.add(line);
 }
}
while (line != null)
```

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/header/header/InputHeaderBlankLines.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/header/header/InputHeaderBlankLines.java]: please create one more test case , with the same content as from issue  (it is good practice ), additionally user reported problem on header that is multiline comment not set of single lines.
The more test cases the better :) .

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/header/RegexpHeaderCheck.java (hunk 2 above)
Comment (kemitix) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/header/RegexpHeaderCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/header/RegexpHeaderCheck.java (hunk 3 above)
Comment (kemitix) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/header/RegexpHeaderCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/header/RegexpHeaderCheck.java (hunk 4 above)
Comment (kemitix) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/header/RegexpHeaderCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java (hunk 5 above)
Comment (kemitix) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/header/AbstractHeaderCheck.java]: Done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/header/header/InputHeaderBlankLines.java (hunk 6 above)
Comment (kemitix) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/header/header/InputHeaderBlankLines.java]: Done

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/header/HeaderCheckTest.java [hunk 7]
```diff
@@ -289,4 +289,19 @@ public void testLoadHeaderFileTwice() throws Exception {
         }
     }
 
+    @Test
+    public void testHeaderIsValidWithBlankLines() throws Exception {
+        final DefaultConfiguration checkConfig = createModuleConfig(HeaderCheck.class);
+        checkConfig.addAttribute("headerFile", getPath("InputHeaderjava.blank-lines.header"));
+        checkConfig.addAttribute("ignoreLines", "");
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/header/HeaderCheckTest.java]: http://checkstyle.sourceforge.net/config_header.html#Header
ignoreLines are empty by default, why mention it here ?

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/header/HeaderCheckTest.java [hunk 8]
```diff
@@ -289,4 +289,19 @@ public void testLoadHeaderFileTwice() throws Exception {
...
+        verify(checkConfig, getPath("InputHeaderBlankLines.java"));
+    }
+
+    @Test
+    public void testHeaderIsValidWithBlankLinesBlockStyle() throws Exception {
+        final DefaultConfiguration checkConfig = createModuleConfig(HeaderCheck.class);
+        checkConfig.addAttribute("headerFile", getPath("InputHeaderjava.blank-lines2.header"));
+        checkConfig.addAttribute("ignoreLines", "");
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/header/HeaderCheckTest.java]: same a above.

**Comment by rnveach** [comment]
@kemitix CI is failing.

**Comment by kemitix** [comment]
@rnveach They look like they might be network issues.
Shippable: `Caused by: java.io.FileNotFoundException: http://maven.apache.org/xsd/decoration-1.0.0.xsd`
Travis: `Downloaded from sonatype: https://oss.sonatype.org/content/repositories/releases/junit/junit-dep/maven-metadata.xml (563 B at 1.4 kB/s)
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.`
Is there a way to rerun those builds?

**Comment by romani** [comment]
@kemitix , thanks a lot for your contribution.

**Pull request merged**